runtime/testy/runtime.py

```diff
@@ -126,25 +128,51 @@ def checkTimedOut(func, *args):
 
     if testName != "":
       testName += "_"
-    setupFunc = getattr(studentCode, testName + "setup")
-    mainFunc = getattr(studentCode, testName + "main")
+    try:
+      setupFunc = getattr(studentCode, testName + "setup")
+    except AttributeError:
+      raise RuntimeError("Student code failed to define '{}'".format(test_name + "setup"))
+    try:
+      mainFunc = getattr(studentCode, testName + "main")
+    except AttributeError:
+      raise RuntimeError("Student code failed to define '{}'".format(test_name + "main"))
+
+    ensure_is_function(testName + "setup", setupFunc)
+    ensure_is_function(testName + "main", mainFunc)
+    ensure_not_overridden(studentCode, 'Robot')
+    ensure_not_overridden(studentCode, 'delay')
 
     r = studentAPI.Robot(stateQueue, pipe)
     studentCode.Robot = r
 
     checkTimedOut(setupFunc)
 
-    # TODO: Replace execCount with a value in stateManager
-    execCount = 0
-    while (not terminated) and (maxIter is None or execCount < maxIter):
-      checkTimedOut(mainFunc)
-      nextCall = time.time()
-      nextCall += 1.0/RUNTIME_CONFIG.STUDENT_CODE_HZ.value
-      stateQueue.put([SM_COMMANDS.STUDENT_MAIN_OK, []])
-      time.sleep(max(nextCall - time.time(), 0))
-      execCount += 1
+    exception_cell = [None]
+    clarify_coroutine_warnings(exception_cell)
 
-    badThingsQueue.put(BadThing(sys.exc_info(), "Process Ended", event=BAD_EVENTS.END_EVENT))
+    async def main_loop():
+      execCount = 0
+      while (exception_cell[0] is None) and (maxIter is None or execCount < maxIter):
+        next_call = loop.time() + 1. / RUNTIME_CONFIG.STUDENT_CODE_HZ.value
+        checkTimedOut(mainFunc)
+
+        sleep_time = max(next_call - loop.time(), 0.)
+        stateQueue.put([SM_COMMANDS.STUDENT_MAIN_OK, []])
+        execCount += 1
+        await asyncio.sleep(sleep_time)
+
+      badThingsQueue.put(BadThing(sys.exc_info(), "Process Ended", event=BAD_EVENTS.END_EVENT))
+      if not (maxIter is None or execCount < maxIter):
```

Comment:
In retrospect, I think a condition of `if exception_cell[0] is not None` would be cleaner and possibly more correct.